csharp/src/Apache.Arrow/Schema.cs

```diff
@@ -63,5 +63,68 @@ public int GetFieldIndex(string name, StringComparer comparer = default)
             return _fields.IndexOf(
                 _fields.Single(x => comparer.Compare(x.Name, name) == 0));
         }
+
+        public override bool Equals(object obj)
+        {
+            if (obj == null || !this.GetType().Equals(obj.GetType()))
+            {
+                return false;
+            }
+            return Equals((Schema)obj);
+        }
+
+        public bool Equals(Schema other)
+        {
+            if (this == other)
+            {
+                return true;
+            }
+            var otherFields = other._fields;
+
+            if (this.HasMetadata != other.HasMetadata)
+            {
+                return false;
+            }
+            if (this._fields.Count != otherFields.Count)
+            {
+                return false;
+            }
+            
+            for (int ii = 0; ii < _fields.Count; ii++)
+            {
+                if (!_fields[ii].Equals(otherFields[ii]))
+                {
+                    return false;
+                }
+            }
+
+            if (this.HasMetadata && other.HasMetadata)
```

Comment:
@stephentoub - have you ever seen a type that internally has a `Dictionary` that overrides `Equals`? Is there a "typical" way of comparing and hashing a `Dictionary`?

@pgovind - maybe the question should be: do we need to override `Equals` on `Schema` and `Field`?

Reply:
> have you ever seen a type that internally has a Dictionary that overrides Equals? Is there a "typical" way of comparing and hashing a Dictionary?

Yeah, they come up here and there.  For example, it's not exactly a `Dictionary`, but it's the same idea:
https://github.com/dotnet/corefx/blob/1c802d6d8c06b727c102c00e5eb4224de9608a16/src/System.Net.Http/src/System/Net/Http/Headers/MediaTypeHeaderValue.cs#L113-L124

> Is there a "typical" way of comparing and hashing a Dictionary?

In such cases I often see the collection left out of the hash.